cadcdata/cadcdata/storageinv.py

```diff
@@ -0,0 +1,826 @@
+        if replace and md5_checksum and (file_info.md5sum == md5_checksum):
+            if (file_info.file_type != headers['Content-Type']) or \
+               (file_info.encoding != headers['Content-Encoding']):
+                operation = 'post'
+            else:
+                logger.info(
+                    'Source {} already in the storage inventory'.format(src))
+                return
+
+        urls = self._get_transfer_urls(id, is_get=False)
+        if len(urls) == 0:
+            raise exceptions.HttpException('No URLs available to put data to')
+
+        # get the list of transfer points
+        for url in urls:
+            if operation == 'post':
+                logger.debug('POST to URL {}'.format(url))
+                result = self._cadc_client.post(url, headers=headers)
+                result.raise_for_status()
+                logger.debug('Updated metadata for identifier {}'.format(id))
+                return
+            logger.debug('PUT to URL {}'.format(url))
+            try:
+                start = time.time()
+                with open(src, 'rb') as f:
+                    self._cadc_client.put(url, headers=headers, data=f)
+                duration = time.time() - start
+                stat_info = os.stat(src)
+                logger.info(
+                    ('Successfully uploaded file {} in {}s '
+                     '(avg. speed: {}MB/s)').format(
+                        id, round(duration, 2),
+                        round(stat_info.st_size / 1024 / 1024 / duration, 2)))
+                return
+            except (exceptions.HttpException, socket.timeout) as e:
+                # try a different URL
+                logger.info('WARN: Cannot {} data to {}. Exception: {}'.
+                            format(operation, url, e))
+                logger.warn('Try the next URL')
+                continue
+        raise exceptions.HttpException(
+            'Unable to {} data from any of the available '
+            'URLs'.format(operation))
+
+    def cadcremove(self, id):
+        """
+        Puts a file into the inventory system
+        :param id: unique identifier (URI) for the file in the CADC inventory
+        system
+        """
+
+        # We actually raise an exception here since the web
+        # service will normally respond with a 200 for an
+        # anonymous put, though not provide any endpoints.
+        if self._cadc_client.subject.anon:
+            raise exceptions.UnauthorizedException(
+                'Must be authenticated to remove data')
+
+        urls = self._get_transfer_urls(id, is_get=False)
+        if len(urls) == 0:
+            raise exceptions.NotFoundException(
+                'File not found: {}'.format(id))
+
+        # get the list of transfer points
+        for url in urls:
+            logger.debug(
+                'REMOVE file with identifier {} from URL {}'.format(id, url))
+            try:
+                start = time.time()
```

Comment:
For symmetry, if this operation, which does not involve moving bytes around, is going to be timed, should the `POST` at line 483 also be timed?

Reply:
Yes.